Enable DinD rootless for Customer Testing

smgrow opened the pull request:
The goal of this PR is to enable DinD [rootless](https://docs.docker.com/engine/security/rootless/) for customer testing.  This file was created from the original application-runner-scale-set.yaml.  Key Changes for Docker-in-Docker Rootless Mode:

1.  **Application Name and Namespace:**

- Changed name: arc-runner-set to name: arc-runners-d-in-d.
- Changed namespace: arc-runners to namespace: arc-runners-d-in-d.

2. **Runner Group:**

- Updated runnerGroup to arc-runners-d-in-d.

3. **Environment Variables:**

- Added DOCKER_HOST environment variable with the value tcp://localhost:2375.

4. **Volumes:**

- Added a docker-socket volume with emptyDir: {} to support rootless mode.

5. **Init Container:**

- Added an init container dind-rootless using the Docker-in-Docker rootless image docker:20.10-dind-rootless.
- Configured security context and environment variables for rootless Docker setup.

This should create a new set of action runners that use Docker-in-Docker in rootless mode.

smgrow committed:
Create application-runner-scale-set-dind.yaml

smgrow commented:
@jcantosz Does this look good?  After this I believe I would apply the new YAML config with:
`kubectl apply -f arc-runners-d-in-d.yaml`

Then create a test workflow with something like:

```
jobs:
  test-docker:
    runs-on: [self-hosted, arc-runners-d-in-d]
    steps:
    - name: Checkout code
      uses: actions/checkout@v2
    - name: Run Docker
      run: |
        docker run hello-world

```
...am I headed down the right path?

smgrow committed:
Update application-runner-scale-set-dind.yaml

smgrow commented:
**Summary of Key Differences from my new file with the original:**
- Added comments throughout
- Changed the name and releaseName from arc-runners to arc-runners-d-in-d.
- Updated the runnerGroup from arc-runners to arc-runners-d-in-d.
- Added the DOCKER_HOST environment variable in the runner container to configure Docker-in-Docker.
- Added volumeMounts for the Docker socket in the runner container.
- Added a docker-socket volume in the volumes section.
- Introduced an initContainers section to include dind-rootless for Docker-in-Docker rootless mode.
- Updated the destination namespace from arc-runners to arc-runners-d-in-d.

smgrow commented:
Created Test Workflow File here:

https://github.com/containerization-vgroup/test-actions/pull/1

smgrow commented:
Next Steps:

1.  Apply the configuration using kubectl apply -f arc-runners-d-in-d.yaml.
2.  Merge PR for the Workflow test:  https://github.com/containerization-vgroup/test-actions/pull/1
3.  Ensure that the Docker build and run steps complete successfully, verifying Docker-in-Docker rootless mode.

jcantosz commented on azure-east/arc-system/application-runner-scale-set-dind.yaml:
```diff
@@ -0,0 +1,92 @@
+              emptyDir: {}  # Create an empty directory volume for Docker socket (supports rootless mode)
+            initContainers:
+            - name: dind-rootless  # Name of the init container for DIND rootless
+              image: docker:20.10-dind-rootless  # Docker image for DIND rootless
```
If you run into issues with this, here are some potential areas to look at
- You may need `restartPolicy: always` on init container ([docs link](https://kubernetes.io/docs/concepts/workloads/pods/sidecar-containers/#pod-sidecar-containers)), assuming this runs a daemon. Init containers generally must exist before anything in the "containers" section starts, having the restart policy lets you start the init container before other containers without it needing to exit
- consider using a newer version of the rootless container, `26.1` looks like the latest https://hub.docker.com/_/docker
- just a not from the [Docker docs](https://docs.docker.com/engine/security/rootless/#rootless-docker-in-docker) - "The `docker:<version>-dind-rootless` image runs as a non-root user (UID 1000). However, `--privileged` is required for disabling seccomp, AppArmor, and mount masks." - may need to set additional security contexts or set privileged if there are failures with this

smgrow committed:
Update application-runner-scale-set-dind.yaml

Restart Policy: Added restartPolicy: Always for the init container to ensure it runs continuously if it stops.

Updated Image: Changed the Docker-in-Docker rootless image to the latest version (docker:26.1-dind-rootless).

Privileged Mode: Set privileged: true in the securityContext for the init container as required for running the rootless Docker daemon.
